### ISSUE TITLE
Gradient unable to propagate

### ISSUE THREAD

**Issue by xuezhongcailian** [JooyoungLeeETRI/SCR#5]
Hello, I have a question to ask you. I noticed in your code, when solving the importance map, as shown in the red box in the diagram below, the range of the clipping values is non-differentiable, making the gradient unable to propagate. I'm curious about this part.

![image](https://github.com/JooyoungLeeETRI/SCR/assets/31470730/8164f075-1b54-4872-9a35-65003a4eed46)

**Comment by JooyoungLeeETRI** [comment]
Hi,
I appreciate your interest.

With our source code, the unclipped importance map values(out2 + 0.5) are around 0.5 (the middle of the clipping range) at the initial training stage, so the training works with the normal "clip_by_value" operator which regards the gradients as zeros for the values outside of the min-max range.

If you need a clipping operator that maintains the gradients even for the values outside, 
you can find the "clip_by_value_preserve_gradient" operator at the following link:
https://www.tensorflow.org/probability/api_docs/python/tfp/math/clip_by_value_preserve_gradient
(but I'm not sure if that operator exists for the tensorflow 1.x version)

If you have any further questions, please let me know!

**Comment by JooyoungLeeETRI** [comment]
Hi,
If you have any problem in using the clip_by_value_preserve_gradient() function, please refer to the following function:

~~~
def clip_by_value_preserve_gradient(t, clip_value_min, clip_value_max, name=None):
  with tf.name_scope(name or 'clip_by_value_preserve_gradient'):
    t = tf.convert_to_tensor(t, name='t')
    clip_t = tf.clip_by_value(t, clip_value_min, clip_value_max)
    return t + tf.stop_gradient(clip_t - t)
~~~


This is the open code from the tensorflow-probability package.
(https://github.com/tensorflow/probability/blob/8947615fe5b19cbce00441b2a62b4929bcd2ca15/tensorflow_probability/python/math/numeric.py)

Hope this helps!

**Comment by xuezhongcailian** [comment]
Thank you for your reply. I have another question: why is the sigmoid function not used here?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "JooyoungLeeETRI/SCR"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年3月7日(星期四) 下午5:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [JooyoungLeeETRI/SCR] Gradient unable to propagate (Issue #5)





 
Hi,
 If you have any problem in using the clip_by_value_preserve_gradient() function, please refer to the following function:
 
def clip_by_value_preserve_gradient(t, clip_value_min, clip_value_max, name=None):
 with tf.name_scope(name or 'clip_by_value_preserve_gradient'):
 t = tf.convert_to_tensor(t, name='t')
 clip_t = tf.clip_by_value(t, clip_value_min, clip_value_max)
 return t + tf.stop_gradient(clip_t - t)
 
This is the open code from the tensorflow-probability package.
 
Hope this helps!

**Comment by JooyoungLeeETRI** [comment]
Sometimes, the sigmoid function causes a vanishing gradient when the input values become large or small. After experiencing prolonged training time in different models a few times before, I prefer to use clipping whenever possible.

But please note that I'm not sure if clipping works better all the time.

**Comment by JooyoungLeeETRI** [comment]
I'll close this thread. If you have any further questions, please let me know again!
